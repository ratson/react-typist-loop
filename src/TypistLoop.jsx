import PropTypes from 'prop-types'
import React from 'react'

class TypistLoop extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    interval: PropTypes.number,
  }

  static defaultProps = {
    interval: 1000,
  }

  state = {
    currentIndex: 0,
  }

  componentDidMount() {
    this._isMounted = true
  }

  componentWillUnmount() {
    this._isMounted = false
    if (this._timer) {
      clearTimeout(this._timer)
    }
  }

  onTypingDone = () => {
    this._timer = setTimeout(this.showNext, this.props.interval)
  }

  showNext = () => {
    if (!this._isMounted) {
      return
    }
    const { children } = this.props
    this.setState(({ currentIndex }) => ({
      currentIndex: (currentIndex + 1) % React.Children.count(children),
    }))
  }

  render() {
    const { onTypingDone } = this
    const { currentIndex } = this.state
    const { children } = this.props
    return React.Children.map(
      children,
      (child, i) =>
        i === currentIndex && React.cloneElement(child, { onTypingDone }),
    )
  }
}

export default TypistLoop
