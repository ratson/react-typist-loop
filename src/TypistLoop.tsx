import * as React from "react";

export interface TypistLoopProps {
  interval: number;
}

const TypistLoop: React.FC<TypistLoopProps> = (
  { interval = 1000, children },
) => {
  const [currentIndex, setCurrentIndex] = React.useState<number>(0);
  const [mounted, setMounted] = React.useState<boolean>(false);
  const [timer, setTimer] = React.useState<number>();

  React.useEffect(() => {
    setMounted(true);
    return () => {
      setMounted(false);
      if (timer) clearTimeout(timer);
    };
  }, []);

  const showNext = () => {
    if (!mounted) return;
    setCurrentIndex((currentIndex + 1) % React.Children.count(children));
  };

  const onTypingDone = React.useCallback(() => {
    setTimer(setTimeout(showNext, interval));
  }, []);

  return React.Children.map(
    children,
    (child: any, i) =>
      i === currentIndex &&
      React.cloneElement(child, { onTypingDone }),
  );
};

export default TypistLoop;
