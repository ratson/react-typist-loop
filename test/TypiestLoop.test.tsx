import React from "react";
import renderer from "react-test-renderer";
import Typist from "react-typist";
import TypistLoop from "../src/TypistLoop";

test("renders correctly", () => {
  const App: React.FC = () => (
    <TypistLoop interval={3000}>
      {["Hello World", "Good Morning", "Bye"].map((text) => (
        <Typist key={text} startDelay={1000}>
          {text}
        </Typist>
      ))}
    </TypistLoop>
  );
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
