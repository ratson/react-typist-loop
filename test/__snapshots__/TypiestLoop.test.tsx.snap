// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Typist "
>
  <span
    className="Cursor Cursor--blinking"
  >
    |
  </span>
</div>
`;
